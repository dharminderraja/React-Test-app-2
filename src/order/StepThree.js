import React, { Component } from 'react';
import NoteAdd from '@material-ui/icons/NoteAdd';

class StepThree extends Component {
  constructor(props, context) {
    super(props, context);

    this.state = {
      dishes: [{ dish: '', number: 1 }],
      errors: []
    };
  }

  componentWillMount() {
    const { dishes } = this.props.data || [];
    if (dishes && dishes.length > 0) {
      this.setState({ dishes });
    }
  }

  addDish() {
    let { dishes } = this.state;
    dishes.push({ dish: '', number: 1 });
    this.setState({ dishes });
  }

  onChange(i, property, value) {
    let { dishes } = this.state;

    dishes[i] = dishes[i] || {};
    dishes[i][property] = value;

    this.setState({ dishes });
  }

  validate() {
    let { errors, dishes } = this.state;

    dishes.forEach((dish, i) => {
      errors[i] = errors[i] || {};
      errors[i]['dish'] = !dish['dish'] ? 'Please select a dish' : '';
      errors[i]['number'] = dish['number'] < 1 ? 'Please enter a valid number' : '';
    })
    this.setState({ errors });

    let flag = true;
    errors.forEach(err => {
      if (!Object.keys(err).every(k => !err[k])) {
        flag = false;
      }
    })

    return flag;
  }

  nextStep() {
    const { updateData, nextTab, disableTabs } = this.props;
    const { dishes } = this.state;

    if (this.validate()) {
      updateData({ dishes });
      nextTab();
    } else {
      disableTabs(['tabFour']);
    }
  }

  render() {
    const { prevTab } = this.props;
    const list = this.props.dishes || [];
    const { dishes, errors } = this.state;

    return (
      <div className='step-detail'>
        {dishes.map((item, idx) => (
          <div key={idx} className='pick-dish'>
            <div className='left'>
              <span className='title'>Please select a Dish</span>
              <select className='meal' value={item.dish || ''} onChange={evt => this.onChange(idx, 'dish', evt.target.value)}>
                <option value=''>...</option>
                {(list || []).filter(i => {
                  return !dishes.find(k => k.dish !== item.dish && k.dish === i.name)
                }).map((i, idx) => (<option key={idx} value={i.name}>{i.name}</option>))}
              </select>
              {errors[idx] && errors[idx]['dish'] && <span className='alert-warning'>{errors[idx]['dish']}</span>}
            </div>
            <div className='right'>
              <span className='title'>Please enter no of servings</span>
              <input className='number-meal' value={item.number} onChange={evt => this.onChange(idx, 'number', evt.target.value)} type='number' />
              {errors[idx] && errors[idx]['number'] && <span className='alert-warning'>{errors[idx]['number']}</span>}
            </div>
            <div className='clear' />
          </div>
        ))}

        <div className='add-dish' onClick={() => this.addDish()}><NoteAdd/></div>

        <div className='nav-step'>
          <button className='prev' onClick={() => prevTab()}>Previous</button>
          <button className='next' onClick={() => this.nextStep()}>Next</button>
        </div>
      </div>
    );
  }
}

export default StepThree;
