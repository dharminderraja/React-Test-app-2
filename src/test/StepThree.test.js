import React from 'react';
import ReactDOM from 'react-dom';

import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

import StepThree from '../order/StepThree';

describe('Step Three', () => {
  it('should be defined', () => {
    expect(StepThree).toBeDefined();
  });

  it('should render correctly', () => {
    const component = renderer.create(<StepThree />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should show values correctly', () => {
    const dishes = [
      {
        "id": 1,
        "name": "Chicken Burger",
        "restaurant": "Mc Donalds",
        "availableMeals": ["lunch", "dinner"]
      },
      {
        "id": 2,
        "name": "Ham Burger",
        "restaurant": "Mc Donalds",
        "availableMeals": ["lunch", "dinner"]
      },
      {
        "id": 3,
        "name": "Cheese Burger",
        "restaurant": "Mc Donalds",
        "availableMeals": ["lunch", "dinner"]
      },
      {
        "id": 4,
        "name": "Fries",
        "restaurant": "Mc Donalds",
        "availableMeals": ["lunch", "dinner"]
      }
    ];

    const tree = shallow(
      <StepThree
        dishes={dishes}
        data={{ dishes: [{ dish: 'Chicken Burger', number: 2 }] }}
        prevTab={() => { }} />
    );

    expect(tree.find('.step-detail .meal').children()).toHaveLength(5);
    expect(tree.find('.step-detail .meal').getElement().props.value).toEqual('Chicken Burger');
  });
});