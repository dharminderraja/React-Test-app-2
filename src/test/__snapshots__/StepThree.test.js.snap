// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Step Three should render correctly 1`] = `
<div
  className="step-detail"
>
  <div
    className="pick-dish"
  >
    <div
      className="left"
    >
      <span
        className="title"
      >
        Please select a Dish
      </span>
      <select
        className="meal"
        onChange={[Function]}
        value=""
      >
        <option
          value=""
        >
          ...
        </option>
      </select>
    </div>
    <div
      className="right"
    >
      <span
        className="title"
      >
        Please enter no of servings
      </span>
      <input
        className="number-meal"
        onChange={[Function]}
        type="number"
        value={1}
      />
    </div>
    <div
      className="clear"
    />
  </div>
  <div
    className="add-dish"
    onClick={[Function]}
  >
    <img
      alt="add-icon"
      className="add-icon"
      src="ic-add.svg"
    />
  </div>
  <div
    className="nav-step"
  >
    <button
      className="prev"
      onClick={[Function]}
    >
      Previous
    </button>
    <button
      className="next"
      onClick={[Function]}
    >
      Next
    </button>
  </div>
</div>
`;
